### PR TITLE
extract make_* functions out of make_*_loader

### PR DESCRIPTION
**Pull request by pmeier**
Our current `make_*` functions are thin wrappers around the corresponding `make_*_loader` function. The latter is only needed for the "deprecated" v2 tests. Thus, we invert the logic here and create a standalone `make_*` function that has a corresponding `make_*_loader` function that depends on it.

### CODE REVIEW

File: test/common_utils.py [hunk 1]
```diff
@@ -492,6 +492,34 @@ def get_num_channels(color_space):
...
+    spatial_size,
+    *,
+    color_space="RGB",
+    batch_dims=(),
+    dtype=torch.float32,
+    device="cpu",
+    constant_alpha=True,
+    memory_format=torch.contiguous_format,
+):
+    spatial_size = _parse_spatial_size(spatial_size)
```
Comment (pmeier) [inline test/common_utils.py:505]: We can be more strict here if we like. Currently also supports passing a single integer to be treated as edge size of a square image. I'm ok with only taking two-tuples (or sequences in general). LMK what you prefer.

File: test/common_utils.py [hunk 2]
```diff
@@ -492,6 +492,34 @@ def get_num_channels(color_space):
     return num_channels
 
 
+def make_image(
+    spatial_size,
+    *,
+    color_space="RGB",
+    batch_dims=(),
+    dtype=torch.float32,
```
Comment (NicolasHug) [inline test/common_utils.py:500]: In practice, transforms largely operate on `uint8` images as they come straight from PIL. Should this be the default instead?

File: test/common_utils.py [hunk 3]
```diff
@@ -492,6 +492,34 @@ def get_num_channels(color_space):
...
+    device="cpu",
+    constant_alpha=True,
```
Comment (NicolasHug) [inline test/common_utils.py:502]: Do we ever use that (in the new tests)?

**Review by NicolasHug — COMMENTED** [review]
Thanks Philip, took a quick glance as discussed offline

File: test/common_utils.py (hunk 3 above)
Comment (pmeier) [reply test/common_utils.py:502]: So far, we have been only testing on RGB and this is a setting that only applies to `"GRAYSCALE_ALPHA"` and `"RGBA"`. I remember we put this in for `convert_color_space`, but since we ripped that out later, not sure that we still need it. IMO it depends if we want to keep the generic `color_space` parameter. If so, I would keep that as well or at least hardcode the behavior, to avoid producing unrealistic images.

File: test/common_utils.py (hunk 2 above)
Comment (pmeier) [reply test/common_utils.py:500]: Sure, fine with me.

File: test/common_utils.py (hunk 3 above)
Comment (pmeier) [reply test/common_utils.py:502]: As discussed offline, we'll remove this parameter for now.

File: test/common_utils.py [hunk 4]
```diff
@@ -501,24 +523,25 @@ def make_image_loader(
     constant_alpha=True,
     memory_format=torch.contiguous_format,
 ):
+    if not constant_alpha:
+        raise ValueError("This should never happen")
```
Comment (pmeier) [inline test/common_utils.py:527]: Per comment. This will be removed with all the other "old" v2 functionalites.

File: test/common_utils.py [hunk 5]
```diff
@@ -492,6 +492,28 @@ def get_num_channels(color_space):
     return num_channels
 
 
+def make_image(
+    spatial_size, *, color_space="RGB", batch_dims=(), dtype=None, device="cpu", memory_format=torch.contiguous_format
```
Comment (NicolasHug) [inline test/common_utils.py:496]: Do we really need to call this `spatial_size`? Is `size` not good enough? We use `size` absolutely everywhere.

(We've had discussions in the past about whether `spatial_size` was even relevant for bboxes, and decided to not address due to lack of time.)

File: test/common_utils.py (hunk 1 above)
Comment (NicolasHug) [reply test/common_utils.py:498]: Only introduce it when/if we need it?

File: test/common_utils.py (hunk 4 above)
Comment (NicolasHug) [reply test/common_utils.py:527]: To clarify: we'll remove the entire `make_image_loader` util, right?

File: test/test_transforms_v2_refactored.py [hunk 6]
```diff
@@ -308,36 +308,26 @@ def wrapper(input, *args, **kwargs):
     return wrapper
 
 
-def make_input(input_type, *, dtype=None, device="cpu", spatial_size=(17, 11), mask_type="segmentation", **kwargs):
+def make_input(input_type, *, spatial_size=(17, 11), mask_type="segmentation", **kwargs):
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:311]: Why move `dtype` and `device` to `kwargs` and not `spatial_size`?
I'm starting to wonder whether we need `make_input` at-all. We could parametrize over `make_xyz` instead of parametrizing over the `input_type` parameter.

File: test/test_transforms_v2_refactored.py [hunk 7]
```diff
@@ -308,36 +308,26 @@ def wrapper(input, *args, **kwargs):
...
-            make_mask = make_detection_mask
-            default_dtype = torch.bool
-        else:
-            raise ValueError(f"`mask_type` can be `'segmentation'` or `'detection'`, but got {mask_type}.")
-        input = make_mask(size=spatial_size, dtype=dtype or default_dtype, device=device, **kwargs)
+        make_mask = {
+            "segmentation": make_segmentation_mask,
+            "detection": make_detection_mask,
+        }[mask_type]
+        input = make_mask(spatial_size, **kwargs)
```
Comment (NicolasHug) [inline test/test_transforms_v2_refactored.py:325]: Following up on comment just above, here we're not even leveraging the `mask_type` parameter of `make_input`, which seems to indicate we don't really need `make_input`

File: test/common_utils.py (hunk 4 above)
Comment (pmeier) [reply test/common_utils.py:527]: Yes. Everything postfixed `_loader` as well as all support classes and functions that come with that.

File: test/common_utils.py (hunk 5 above)
Comment (pmeier) [reply test/common_utils.py:496]: > Do we really need to call this `spatial_size`? Is `size` not good enough? We use `size` absolutely everywhere.

That is the term we (before your time) decided to use. For example, the main helper function to extract the value from all our supported objects is

https://github.com/pytorch/vision/blob/2d4484fba0f45637f68adadf5a056a6147642aa4/torchvision/transforms/v2/functional/_meta.py#L117

Our reason for this was twofold:

1. Calling it `image_size` excludes the use case of videos
2. Just calling it `size` can lead to ambiguities for bounding boxes, since for them the value is independent of the actual tensor shape. For all other datapoints `spatial_size` and `.shape` are tightly coupled.

Thus, I've opted to call it `spatial_size` here as well. I'm ok with changing it to a better name if we find one, if we reflect that in the actual API as well.

> decided to not address due to lack of time

That is not what I remember. We left it as is because all alternatives were way more complex. Just as a quick example: how would you call a dispatcher with a bounding box if the spatial size (or whatever we might call it) is not stored on the object itself?

File: test/test_transforms_v2_refactored.py (hunk 6 above)
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:311]: > Why move `dtype` and `device` to `kwargs` and not `spatial_size`?

Because `make_input` currently has a default value for `spatial_size`, but the lower level functions do not. Will change that.

File: test/test_transforms_v2_refactored.py (hunk 7 above)
Comment (pmeier) [reply test/test_transforms_v2_refactored.py:325]: I don't understand. We are using `mask_type` to select the right function and store that as `make_mask`. What do you mean by 

> here we're not even leveraging the `mask_type` parameter

?

Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py:325]: > of make_input

File: test/common_utils.py (hunk 5 above)
Comment (NicolasHug) [reply test/common_utils.py:496]: You can see https://github.com/pytorch/vision/issues/7217 and the links there

Comment (NicolasHug) [reply test/common_utils.py]: > Just calling it size can lead to ambiguities for bounding boxes

But how is this ambiguity resolved now that everything is called `spatial_size`? `spatial_size` refers to the size of an image and for a bbox it's the `spatial_size` of its image. This is literally the same situation as when `size` was used for everything?

I understand the need to distinguish between the size of the image of a bbox and the size of the bbox. But when everythign is called spatial_size, I don't see how we solved our issue

File: test/test_transforms_v2_refactored.py (hunk 7 above)
Comment (NicolasHug) [reply test/test_transforms_v2_refactored.py]: sorry for my confusion, I didn't realize we were actuall in `make_input`

File: test/common_utils.py (hunk 5 above)
Comment (pmeier) [reply test/common_utils.py]: The term "size" in PyTorch land is equivalent to the shape

```pycon
>>> t = torch.empty(2, 3, 4)
>>> t.shape
torch.Size([2, 3, 4, 5])
>>> t.size()
torch.Size([2, 3, 4, 5])
```

Meaning, calling everything `spatial_size` removes the ambiguity that we necessarily talk about the `.shape` of the tensor. It does not mean that `spatial_size` cannot be coupled with the shape, but it doesn't have to be.

Comment (NicolasHug) [reply test/common_utils.py]: The term "size" in torchvision (e.g. in `Resize()` or the past `get_image_size()` etc.) refers to `H, W`. It's strongly tied to `.shape` and by extension to `.size()`, but it's not exactly that. There is no ambiguity about the use of "size" in torchvision.

I also don't think this argument stands considering we wouldn't have used the term `spatial_size` if it weren't for bboxes.

**Comment by pmeier** [comment]
Regarding removing `make_input`: I've played around with it and it works mostly fine. The only thing that will be harder is the datapoint / simple tensor / PIL Image handling. Currently this is logic that we have inside `make_input`:

https://github.com/pytorch/vision/blob/2d4484fba0f45637f68adadf5a056a6147642aa4/test/test_transforms_v2_refactored.py#L312-L317

Since we need this somehow, one possibility is to add two new functions `make_image_tensor` and `make_image_pil` that are thin wrappers around `make_image`, which produces a datapoint in line with all the other "regular" `make_*` functions.

Another option would be to add a `return_type` parameter to `make_image` and inside our tests just `functools.partial(make_image, return_type=...)`. 

Regardless of which option we choose, if we remove `make_input` we are no longer parametrizing over `input_type`, but rather something like

```python
@pytest.mark.parametrize(
    ("kernel", "make_input"),
    [
        # TODO: add simple tensor / PIL image handling
        (F.resize_image_tensor, make_image),
        (F.resize_bounding_box, make_bounding_box),
        (F.resize_mask, make_segmentation_mask),
        (F.resize_video, make_video),
    ],
)
```

That should work fine in general, but I already found one instance in the tests ported so far, that branch on the `input_type`:

https://github.com/pytorch/vision/blob/2d4484fba0f45637f68adadf5a056a6147642aa4/test/test_transforms_v2_refactored.py#L748-L753

If we go for the standalone functions, we can still do this although it looks a little quirky

```python
if make_input is make_image_tensor and interpolation is transforms.InterpolationMode.NEAREST_EXACT:
    ...
```

However, if we use `functools.partial` that is not possible.

Admittedly, the branching above is uncommon and I don't expect significantly more of this pattern in the future. Still, it begs the question if `make_input` in fact has a reason to exist. Thoughts?

**Comment by NicolasHug** [comment]
Thanks for investigating Philip

> Admittedly, the branching above is uncommon and I don't expect significantly more of this pattern in the future

If that's the case then I guess we can safely get rid of it

**Comment by pmeier** [comment]
Re `size` vs. `spatial_size`: we had a long offline conversation about and came to the following conclusions:

- `size` is a valid term for bounding boxes. It describes the size of the bounding box, i.e.
  - (box[2] - box[0], box[3] - box[1]) for XYXY
  - (W, H) for XYWH and CXCYWH
  
  It is unrelated to the tensor shape. That is different from all the other datapoints that we currently have, where the size is directly encoded in the shape.
- We still need a term to describe the size of the image, video, mask as well as the corresponding "image size" of a bounding box. Out of lack for a better term, we agreed to keep `spatial_size` for now.
- The `make_*` functions don't need to know about `spatial_size` in general, but only `make_bounding_box` needs it. We can add a `size` attribute to it and sample a `spatial_size` around it if none is given. Meaning, all `make_*` functions will have a `size` attribute and that is what we are going to use going forward.
  - Note that `size` in `make_bounding_box` diverges from the `size` in the other functions. Since it describes the size of the bounding box, the `spatial_size` will be larger. Meaning,

    ```python
    image = make_image(size=(17, 11))
    box = make_bounding_box(size=(17, 11))
    ```
    will no longer produce a valid sample, since the spatial size of the image is (17, 11), while the spatial size of the bounding box is not. For this use case, one needs to now do
    ```python
    image = make_image(size=(17, 11))
    box = make_bounding_box(spatial_size=(17, 11))
    ```

File: test/common_utils.py (hunk 5 above)
Comment (pmeier) [reply test/common_utils.py]: Summary after further offline discussion in https://github.com/pytorch/vision/pull/7717#issuecomment-1621519282.

File: test/test_transforms_v2_refactored.py [hunk 8]
```diff
@@ -773,7 +761,10 @@ def test_transform_unknown_size_error(self):
         [torch.Tensor, PIL.Image.Image, datapoints.Image, datapoints.BoundingBox, datapoints.Mask, datapoints.Video],
     )
     def test_noop(self, size, input_type):
-        input = make_input(input_type, spatial_size=self.INPUT_SIZE)
+        input = make_input(
+            input_type,
+            **{"spatial_size" if issubclass(input_type, datapoints.BoundingBox) else "size": self.INPUT_SIZE},
+        )
```
Comment (pmeier) [inline test/test_transforms_v2_refactored.py:767]: I've missed this in my assessment in https://github.com/pytorch/vision/pull/7717#issuecomment-1621236156. The issue here is that we want to test if the resizing is a no-op. Since `resize` acts on the spatial size of a bounding box, we need to make sure to use that here.

Comment (pmeier) [reply test/test_transforms_v2_refactored.py:767]: If we don't want to do that, we could also do 

```python
output = F.resize(input, size=F.get_spatial_size(input), antialias=True)
```

below. Or being more explicit:

```python
output = F.resize(
    input, 
    size=input.spatial_size if isinstance(input, datapoints.BoundingBox) else self.INPUT_SIZE, 
    antialias=True,
)
```

File: test/transforms_v2_kernel_infos.py [hunk 9]
```diff
@@ -260,6 +261,9 @@ def reference_inputs_convert_format_bounding_box():
         reference_fn=reference_convert_format_bounding_box,
         reference_inputs_fn=reference_inputs_convert_format_bounding_box,
         logs_usage=True,
+        closeness_kwargs={
+            (("TestKernels", "test_against_reference"), torch.int64, "cpu"): dict(atol=1, rtol=0),
+        },
```
Comment (pmeier) [inline test/transforms_v2_kernel_infos.py:266]: I guess we were lucky before that we never hit that. Converting to or from CXCYWH with integer dtypes can lead to rounding errors. Now that we have refactored `make_bounding_box`, we are seeing two of those errors in this test.

**Comment by pmeier** [comment]
The diff in this PR grew pretty large so here is a summary of the changes:

- Factor out the make_* functions out of the corresponding make_*_loader
  - Since we still have v2 tests that rely on the loader architecture, these functions need to stay for now. However, if we are done porting, we can remove them as well as all supporting classes and functions.
- Remove the `"random"` option for the size of the generated inputs. This amounts for most of the changes in `test/transforms_v2_kernel_infos.py`
- Add a `size` parameter to `make_bounding_box`
- Remove `make_input` in favor of parametrizing over the low-level functions

File: test/test_transforms_v2_refactored.py (hunk 8 above)
Comment (pmeier) [reply test/test_transforms_v2_refactored.py]: I went for 

```python
output = F.resize(input, size=F.get_spatial_size(input), antialias=True)
```

File: test/common_utils.py [hunk 10]
```diff
@@ -589,76 +624,90 @@ class BoundingBoxLoader(TensorLoader):
...
+        # We cannot use torch.randint directly here, because it only allows integer scalars as values for low and high.
+        # However, if we have batch_dims, we need tensors as limits.
+        return torch.stack([torch.randint(max_value - v, ()) for v in values.flatten().tolist()]).reshape(values.shape)
+
+    if isinstance(format, str):
+        format = datapoints.BoundingBoxFormat[format]
+
+    if spatial_size is None:
+        if size is None:
+            spatial_size = DEFAULT_SIZE
```
Comment (NicolasHug) [inline test/common_utils.py:646]: I'm a bit confused by this logic here and mostly have 2 Qs:

- Do we ever call `make_bounding_box()` with both size and spatial_size to None?
- If we do, shouldn't the logic be: 1) set `size` to `DEFAULT_SIZE` (for consistency with the rest of the helpers) and 2) generate `spatial_size` accordingly?

Comment (NicolasHug) [reply test/common_utils.py:646]: > My logic here was that if we want to generate an arbitrary sample, we can just do

OK why not - maybe just add that as a comment then

**Review by NicolasHug — APPROVED** [review]
Thanks Philip

**Comment by pmeier** [comment]
Test failures stem likely from a bad reference function. I can reproduce the same error on `main` when using the bounding box created by this PR. However, the failing test is actually a correctness test that should have been removed in #7713, but was forgotten. I acted on that in 519f0fad9af28e8dc95206df449ffb40cd5ac104 and thus there is no need to fix any test.
